pkg/sentry/devices/memdev/kmsg.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memdev
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/safemem"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+	"gvisor.dev/gvisor/pkg/syserror"
+	"gvisor.dev/gvisor/pkg/usermem"
+	"gvisor.dev/gvisor/pkg/waiter"
+)
+
+const (
+	// BufferEntryMax is the maximum number of entries retained by kmsg before new entries overwrite old ones.
+	BufferEntryMax = 512
+
+	// BufferMax is the maximum size of an individual entry, in bytes.
+	BufferMax = 1024
+)
+
+const kmsgDevMinor = 11
+
+// kmsgDevice implements vfs.Device for /dev/kmsg.
+type kmsgDevice struct{}
+
+func (kmsgDevice) Open(ctx context.Context, mnt *vfs.Mount, vfsd *vfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
+	fd := &kmsgFD{}
+	if err := fd.vfsfd.Init(fd, opts.Flags, mnt, vfsd, &vfs.FileDescriptionOptions{
+		UseDentryMetadata: true,
+	}); err != nil {
+		return nil, err
+	}
+	return &fd.vfsfd, nil
+}
+
+// kmsgFD implements vfs.FileDescriptionImpl for /dev/kmsg.
+type kmsgFD struct {
+	vfsfd vfs.FileDescription
+	vfs.FileDescriptionDefaultImpl
+	vfs.DentryMetadataFileDescriptionImpl
+	vfs.NoLockFD
+
+	// buffer is a circular buffer that store user's write to kmsg.
+	m            sync.Mutex
+	buffer       [BufferEntryMax]*bytes.Buffer
+	readIndex    int
+	writeIndex   int
+	bufferLength int
```

Comment:
A comment on `bufferLength` may be helpful:

```
// bufferLength is the number of unread entries in the buffer
```

The meaning of "length" is otherwise ambiguous.